File: sdk/node/src/testing.ts
/**
 * Testing utilities for CAP Node SDK.
 * Allows testing handlers without running NATS or Redis.
 */
import type { Msg, NatsConnection, Subscription, SubscriptionOptions } from "nats";
import { z } from "zod";
import { Agent, InMemoryBlobStore, AgentOptions, Context } from "./runtime";
import { loadRoot, DEFAULT_PROTOCOL_VERSION } from "./protos";
import { encodeDeterministic } from "./codec";

/** A message published on the mock bus. */
export interface PublishedMsg {
  subject: string;
  data: Uint8Array;
}

/**
 * MockNatsConnection simulates a NATS connection for testing.
 * Tracks published messages and subscription callbacks.
 */
export class MockNatsConnection {
  published: PublishedMsg[] = [];
  subscriptions: Map<
    string,
    NonNullable<SubscriptionOptions["callback"]>
  > = new Map();

  async publish(subject: string, data: Uint8Array): Promise<void> {
    this.published.push({ subject, data });
  }

  subscribe(subject: string, opts: SubscriptionOptions = {}): Subscription {
    if (opts.callback) {
      this.subscriptions.set(subject, opts.callback);
    }
    return {
      drain: async () => {},
      unsubscribe: () => {},
    } as unknown as Subscription;
  }

  deliver(subject: string, data: Uint8Array): void {
    const callback = this.subscriptions.get(subject);
    if (!callback) {
      throw new Error(`no callback subscription found for subject ${subject}`);
    }
    callback(null, { subject, data } as unknown as Msg);
  }

  async drain(): Promise<void> {
    return;
  }
}

/**
 * Creates an Agent pre-configured with MockNatsConnection + InMemoryBlobStore.
 * No NATS or Redis required.
 */
export function createTestAgent(
  options: Partial<AgentOptions> = {}
): { agent: Agent; bus: MockNatsConnection; store: InMemoryBlobStore } {
  const bus = new MockNatsConnection();
  const store = new InMemoryBlobStore();
  const agent = new Agent({
    store,
    connectFn: async () => bus as unknown as NatsConnection,
    senderId: options.senderId ?? "test-worker",
    retries: options.retries ?? 0,
    ...options,
  });
  return { agent, bus, store };
}

async function waitForPublish(bus: MockNatsConnection, initialCount: number, timeoutMs = 1000): Promise<void> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    if (bus.published.length > initialCount) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 5));
  }
  throw new Error("timeout waiting for handler to publish result");
}

/**
 * Runs a single handler invocation and returns the decoded JobResult.
 * No NATS or Redis required.
 *
 * @param handler - The handler function to test
 * @param input - The input data to pass to the handler
 * @param options - Optional: inputSchema, outputSchema, senderId, retries
 * @returns The decoded JobResult from the handler
 */
export async function testHandler<TIn, TOut>(
  handler: (ctx: Context, data: TIn) => Promise<TOut> | TOut,
  input: TIn,
  options: {
    inputSchema?: z.ZodType<TIn>;
    outputSchema?: z.ZodType<TOut>;
    senderId?: string;
    retries?: number;
    jobId?: string;
    topic?: string;
  } = {}
): Promise<{ status: number; jobId: string; resultPtr: string; errorMessage: string; workerId: string }> {
  const { agent, bus, store } = createTestAgent({
    senderId: options.senderId,
    retries: options.retries,
  });

  const topic = options.topic ?? "job.test";
  const jobId = options.jobId ?? "test-job-1";

  if (options.inputSchema) {
    agent.job<TIn, TOut>(topic, options.inputSchema, handler, {
      outputSchema: options.outputSchema,
    });
  } else {
    agent.job<TIn, TOut>(topic, handler, {
      outputSchema: options.outputSchema,
    });
  }

  await agent.start();

  const ctxKey = `ctx:${jobId}`;
  await store.set(ctxKey, Buffer.from(JSON.stringify(input), "utf-8"));

  const root = await loadRoot();
  const BusPacket = root.lookupType("cordum.agent.v1.BusPacket");

  const packet = BusPacket.fromObject({
    traceId: "test-trace",
    senderId: "test-client",
    protocolVersion: DEFAULT_PROTOCOL_VERSION,
    createdAt: { seconds: Math.floor(Date.now() / 1000), nanos: 0 },
    jobRequest: { jobId, topic, contextPtr: `redis://${ctxKey}` },
  });
  const payload = encodeDeterministic(BusPacket, packet);

  const initialCount = bus.published.length;
  bus.deliver(topic, payload);

  await waitForPublish(bus, initialCount);

  const resultPacket = BusPacket.decode(
    bus.published[bus.published.length - 1].data
  ) as unknown as { jobResult?: JobResultView };
  const jr = resultPacket.jobResult;
  if (!jr) {
    throw new Error("worker did not publish a JobResult");
  }

  await agent.close();

  return {
    status: jr.status,
    jobId: jr.jobId,
    resultPtr: jr.resultPtr ?? "",
    errorMessage: jr.errorMessage ?? "",
    workerId: jr.workerId ?? "",
  };
}

interface JobResultView {
  status: number;
  jobId: string;
  resultPtr?: string;
  errorMessage?: string;
  workerId?: string;
}

File: sdk/python/cap/testing.py
"""Testing utilities for CAP Python SDK.

Allows testing handlers without running NATS or Redis.
"""

import asyncio
import json
from typing import Any, Awaitable, Callable, Dict, List, Optional, Tuple, Type, Union

from cap.constants import DEFAULT_PROTOCOL_VERSION
from cap.pb.cordum.agent.v1 import buspacket_pb2, job_pb2
from cap.runtime import Agent, Context, InMemoryBlobStore


class MockNATS:
    """In-memory mock that replaces a real NATS connection for testing."""

    def __init__(self) -> None:
        self.subscriptions: Dict[str, Any] = {}
        self.published: "asyncio.Queue[Tuple[str, bytes]]" = asyncio.Queue()

    async def publish(self, subject: str, data: bytes) -> None:
        await self.published.put((subject, data))

    async def subscribe(self, subject: str, queue: str, cb: Any) -> None:
        self.subscriptions[subject] = cb

    async def connect(self, servers: str, name: str) -> "MockNATS":
        return self

    async def drain(self) -> None:
        return None


def create_test_agent(
    *,
    sender_id: str = "test-worker",
    retries: int = 0,
    **kwargs: Any,
) -> Tuple[Agent, MockNATS, InMemoryBlobStore]:
    """Create an Agent pre-configured with MockNATS + InMemoryBlobStore.

    Returns (agent, mock_nats, store). No NATS or Redis required.
    """
    mock = MockNATS()
    store = InMemoryBlobStore()
    agent = Agent(
        store=store,
        connect_fn=mock.connect,
        sender_id=sender_id,
        retries=retries,
        **kwargs,
    )
    return agent, mock, store


async def run_handler(
    handler: Callable[..., Union[Awaitable[Any], Any]],
    input_data: Any,
    *,
    input_model: Optional[Type[Any]] = None,
    output_model: Optional[Type[Any]] = None,
    job_id: str = "test-job-1",
    topic: str = "job.test",
    sender_id: str = "test-worker",
    retries: int = 0,
) -> job_pb2.JobResult:
    """Run a single handler invocation and return the JobResult.

    No NATS or Redis required.

    Args:
        handler: The handler function to test.
        input_data: The input data (will be JSON-serialized into the blob store).
        input_model: Optional Pydantic model for input validation.
        output_model: Optional Pydantic model for output validation.
        job_id: Job ID for the test request.
        topic: Topic for the test request.
        sender_id: Worker sender ID.
        retries: Number of retries on handler failure.

    Returns:
        The JobResult produced by the handler.
    """
    agent, mock, store = create_test_agent(sender_id=sender_id, retries=retries)

    @agent.job(topic, input_model=input_model, output_model=output_model)
    async def _wrapper(ctx: Context, data: Any) -> Any:
        result = handler(ctx, data)
        if asyncio.iscoroutine(result):
            return await result
        return result

    await agent.start()

    ctx_key = f"ctx:{job_id}"
    await store.set(ctx_key, json.dumps(input_data).encode("utf-8"))

    req = job_pb2.JobRequest(
        job_id=job_id,
        topic=topic,
        context_ptr=f"redis://{ctx_key}",
    )
    packet = buspacket_pb2.BusPacket()
    packet.trace_id = "test-trace"
    packet.sender_id = "test-client"
    packet.protocol_version = DEFAULT_PROTOCOL_VERSION
    packet.job_request.CopyFrom(req)

    cb = mock.subscriptions.get(topic)
    if cb is None:
        raise RuntimeError(f"no subscription found for topic {topic}")

    msg = type("Msg", (object,), {"data": packet.SerializeToString(deterministic=True)})()
    await cb(msg)

    subject, payload = await asyncio.wait_for(mock.published.get(), timeout=5.0)

    result_packet = buspacket_pb2.BusPacket()
    result_packet.ParseFromString(payload)

    await agent.close()

    return result_packet.job_result
